internal/store/interfaces.go

```diff
@@ -47,12 +21,52 @@ type Route interface {
 	ID() string
 }
 
-// Store is used for persisting and querying gateways and routes
+type GatewayRecord struct {
+	ID   core.GatewayID
+	Data []byte
+}
+
+type RouteRecord struct {
+	ID   string
+	Data []byte
+}
+
+// Backend is used for persisting and querying gateways and routes
+type Backend interface {
+	GetGateway(ctx context.Context, id core.GatewayID) ([]byte, error)
+	GetGateways(ctx context.Context) ([][]byte, error)
+	DeleteGateway(ctx context.Context, id core.GatewayID) error
+	UpsertGateways(ctx context.Context, gateways ...GatewayRecord) error
+	GetRoute(ctx context.Context, id string) ([]byte, error)
+	GetRoutes(ctx context.Context) ([][]byte, error)
+	DeleteRoute(ctx context.Context, id string) error
+	UpsertRoutes(ctx context.Context, routes ...RouteRecord) error
+}
+
+type Marshaler interface {
+	UnmarshalRoute(data []byte) (Route, error)
+	MarshalRoute(Route) ([]byte, error)
+	UnmarshalGateway(data []byte) (Gateway, error)
+	MarshalGateway(Gateway) ([]byte, error)
+}
+
+type Binder interface {
+	Bind(ctx context.Context, gateway Gateway, route Route) (bool, error)
+	Unbind(ctx context.Context, gateway Gateway, route Route) (bool, error)
+}
+
+type StatusUpdater interface {
+	UpdateGatewayStatusOnSync(ctx context.Context, gateway Gateway, sync func() (bool, error)) error
+	UpdateRouteStatus(ctx context.Context, route Route) error
+}
+
 type Store interface {
```

Comment:
This is really only used for mocking/testing -- other than mocks, really only the concrete `*store` impllementation should be used.